engine/src/main/java/org/hibernate/validator/constraintvalidation/HibernateCrossParameterConstraintValidatorContext.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.constraintvalidation;
+
+import java.util.List;
+
+import javax.validation.ConstraintValidatorContext;
+
+import org.hibernate.validator.Incubating;
+
+/**
+ * A custom {@link ConstraintValidatorContext} which provides aditional functionality for parameter validation contexts.
+ *
+ * @author Marko Bekhta
+ * @since 6.1.0
+ */
+@Incubating
+public interface HibernateCrossParameterConstraintValidatorContext extends HibernateConstraintValidatorContext {
```

Comment:
I agree that this change generated a lot of changes. Why I wanted to apply it is because I didn't quite like the part where we pass the parameter names all the time and sometimes they are `null` (in all cases except the parameter validation). and I just wanted to get rid of that. We could add another constructor without the parameter names argument and use it in other places, but still we will be doing the checks in this builder. 
As a side effect of this - I wasn't happy how we handled things in `ParameterScriptAssertValidator` where we first check if we can unwrap to `HibernateConstraintValidatorContext` but then just do the cast to `ConstraintValidatorContextImpl`, and that's why I've added this additional interface. 
With all that said :) I definitely would like to clean things up in `ParameterScriptAssertValidator` and as for the rest I'll trust your decision.